Quinlan-C4.5-Release8_and_python_interface_for_EBP/Doc/c4.5.1
.EN
.TH C4.5 1
.SH NAME
.PP
c4.5 \- form a decision tree from a file of examples
.SH SYNOPSIS
.PP
.B c4.5
[ \fB-f\fR filestem ]
[ \fB-u\fR ]
[ \fB-s\fR ]
[ \fB-p\fR ]
[ \fB-v\fR verb ]
[ \fB-t\fR trials ]
   [ \fB-w\fR wsize ]
[ \fB-i\fR incr ]
[ \fB-g\fR ]
[ \fB-m\fR minobjs ]
[ \fB-c\fR cf ]
.SH DESCRIPTION
.PP
.I C4.5
is a program for inducing classification rules in the form
of decision trees from a set of given examples.
.PP
All files read and written by C4.5 are of the form
.I filestem.ext
where
.I filestem
is a file name stem that identifies the induction task and
.I ext
is an extension that defines the type of file.
The program expects to find at least 
two files: a
.B names file
.I filestem.names
defining class, attribute and attribute value names, and a
.B data file
.I filestem.data
containing a set of objects, each of which is described by its
values of each of the attributes and its class.
.PP
The program can generate trees
in two ways.  In
.I batch
mode (the default), the program generates a single tree
using all the available data.
In
.I iterative
mode,
the program starts with a randomly-selected subset of the
data (the
.I window),
generates a trial decision tree, adds some misclassified
objects, and continues until the trial decision tree
correctly classifies all objects not in the window or
until it appears that no progress is being made.
Since iterative mode starts with a randomly-selected subset,
multiple trials with the same data can be used to generate
more than one tree.
.PP
All trees generated in the process are saved in
.I filestem.unpruned.
After each tree is generated, it is
.I pruned
in an attempt to simplify it.
The `best' pruned tree (selected by the program if more there is
more than one trial)
is saved in machine-readable form in
.I filestem.tree.
.PP
All trees produced, both pre- and post-simplification, are evaluated
on the training data.  If required, they can also be evaluated
on unseen data in file
.I filestem.test.

.SH FILE FORMATS
The
.B names file
.I filestem.names
is a series of entries defining names of attributes,
attribute values and classes.  The file is free-format
with the exception that the vertical bar `|' causes the
remainder of that line to be ignored.
Each entry is terminated by a period which may be
omitted if it is the last character of a line.
.PP
The file
commences with the names of the classes, separated by
commas and terminated with a period.  Each name consists of
a string of characters that does not include comma, question mark
or colon (unless preceded by a backslash).  A period may be
embedded in a name provided it is not followed by a space.
Embedded spaces are also permitted but multiple whitespace is
replaced by a single space.
The rest of the file consists of a single entry for each
attribute.  An attribute entry begins with the attribute name
followed by a colon, and then either the word `ignore' (indicating
that this attribute should not be used), the word `continuous'
(indicating that the attribute has real values),
the word `discrete' followed by an integer
.I n
(indicating that the program should assemble
a list of up to
.I n
possible values), or a list
of all possible discrete values separated by commas.  (The latter
form for discrete attributes is recommended as it
enables input to be checked.)  Each
entry is terminated with a period (but see above).
.PP
The
.B data file
.I filestem.data
contains one line per object.  Each line contains
the values of the attributes in order followed by the
object's class, with all entries separated by commas.
The rules for valid names in the
.B names file
also hold for the names in the
.B data file.
An unknown value of an attribute is indicated by a
question mark `?'.
If a 
.B test file
.I filestem.test
is used, it has the same format as the data file.

.SH OPTIONS
Options and their meanings are:
.PP
.TP 12
.BI \-f filestem\^
Specify the filename stem (default
.B DF)
.TP
.B \-u
Evaluate trees produced on unseen cases in file 
.I filestem.test.
.TP
.B \-s
Force `subsetting' of all tests based on discrete attributes
with more than two values.  C4.5 will construct a test with
a subset of values associated with each branch.
.TP
.B \-p
Probabilistic thresholds used for continuous attributes (see Quinlan, 1987a).
.TP
.BI \-t trials\^
Set iterative mode with specified number of trials.
.TP
.BI \-v verb\^
Set the verbosity level [0-3] (default 0).
This option generates more voluminous output that may help to
explain what the program is doing (but don't count on it);
see the manual entry for
.I verbose.
.PP
The following options are also available but need not
be used except for experimentation with tree construction:
.TP 12
.BI \-w wsize\^
Set the size of the initial window
(default is the maximum of 20 percent and twice the square
root of the number of data objects).
.TP
.BI \-i incr\^
Set the maximum number of objects that can be
added to the window at each iteration
(default is 20 percent of the initial window size).
.TP
.B \-g
Use the gain criterion to select tests.  The default
uses the gain ratio criterion.
.TP
.BI \-m minobjs\^
In all tests, at least two branches must contain a minimum number
of objects (default 2).  This option allows the minimum
number to be altered.
.TP
.BI \-c cf\^
Set the pruning confidence level (default 25%).
.SH FILES
.PP
.in 8
c4.5
.br
filestem.data
.br
filestem.names
.br
filestem.unpruned  (unpruned trees)
.br
filestem.tree   (final decision tree)
.br
filestem.test   (unseen data)
.in 0
.PP
.SH SEE ALSO
.PP
consult(1)
.PP
.SH BUGS


Quinlan-C4.5-Release8_and_python_interface_for_EBP/Src/quinlan-src/c4.5.c
/*************************************************************************/
/*									 */
/*	Main routine, c4.5						 */
/*	------------------						 */
/*									 */
/*************************************************************************/


#include "defns.i"
#include "types.i"
#include <stdlib.h>
// 下面是为了消除警告所增加的声明
void PrintHeader();
int getopt();
void GetData();
void OneTree();
void GetNames();
void SoftenThresh();
void Evaluate();
void SaveTree();
void PrintTree();


    /*  External data, described in extern.i  */

short		MaxAtt, MaxClass, MaxDiscrVal = 2;

ItemNo		MaxItem;

Description	*Item;

DiscrValue	*MaxAttVal;

char		*SpecialStatus;

String		*ClassName,
		*AttName,
		**AttValName,
		// FileName = "vote";
        FileName;

short		VERBOSITY = 0,
		TRIALS    = 10;

Boolean		GAINRATIO  = true,
		SUBSET     = false,
		BATCH      = true,
		UNSEENS    = false,
		PROBTHRESH = false;

ItemNo		MINOBJS   = 2,
		WINDOW    = 0,
		INCREMENT = 0;

float		CF = 0.25;

Tree		*Pruned;

Boolean		AllKnown = true;


int main(Argc, Argv)
/*  ----  */
    int Argc;
    char *Argv[];
{

    int o;
    extern char *optarg;
    extern int optind;
    Boolean FirstTime=true;
    short Best, BestTree();

    PrintHeader("decision tree generator");

    /*  Process options  */

    while ( (o = getopt(Argc, Argv, "f:bupv:t:w:i:gsm:c:")) != EOF )
    {
	if ( FirstTime )
	{
	    printf("\n    Options:\n");
	    FirstTime = false;
	}

	switch (o)
	{
	case 'f':   FileName = optarg;
		    printf("\tFile stem <%s>\n", FileName);
		    break;
	case 'b':   BATCH = true;
		    printf("\tWindowing disabled (now the default)\n");
		    break;
	case 'u':   UNSEENS = true;
		    printf("\tTrees evaluated on unseen cases\n");
		    break;
	case 'p':   PROBTHRESH = true;
		    printf("\tProbability thresholds used\n");
		    break;
	case 'v':   VERBOSITY = atoi(optarg);
		    printf("\tVerbosity level %d\n", VERBOSITY);
		    break;
	case 't':   TRIALS = atoi(optarg);
		    printf("\tWindowing enabled with %d trials\n", TRIALS);
		    Check(TRIALS, 1, 10000);
		    BATCH = false;
		    break;
	case 'w':   WINDOW = atoi(optarg);
		    printf("\tInitial window size of %d items\n", WINDOW);
		    Check(WINDOW, 1, 1000000);
		    BATCH = false;
		    break;
	case 'i':   INCREMENT = atoi(optarg);
		    printf("\tMaximum window increment of %d items\n",
			   INCREMENT);
		    Check(INCREMENT, 1, 1000000);
		    BATCH = false;
		    break;
	case 'g':   GAINRATIO = false;
		    printf("\tGain criterion used\n");
		    break;
	case 's':   SUBSET = true;
		    printf("\tTests on discrete attribute groups\n");
		    break;
	case 'm':   MINOBJS = atoi(optarg);
		    printf("\tSensible test requires 2 branches with >=%d cases\n",
			    MINOBJS);
		    Check(MINOBJS, 1, 1000000);
		    break;
	case 'c':   CF = atof(optarg);
		    printf("\tPruning confidence level %g%%\n", CF);
		    Check(CF, Epsilon, 100);
		    CF /= 100;
		    break;
            case '?':FileName=Argv[1];
	// case '?':   printf("unrecognised option\n");
	// 	    exit(1);
	}
    }

    /*  Initialise  */

    GetNames();
    GetData(".data");
    printf("\nRead %d cases (%d attributes) from %s.data\n",
	   MaxItem+1, MaxAtt+1, FileName);

    /*  Build decision trees  */
    printf("BATCH=%d\n",BATCH);
    if ( BATCH )//默认进入if中执行
    {
	TRIALS = 1;
	OneTree();
	Best = 0;
    }
    else
    {
        printf("进入BestTree\n");
	Best = BestTree();
    }

    /*  Soften thresholds in best tree  */

    if ( PROBTHRESH )
    {
	printf("Softening thresholds");
	if ( ! BATCH ) printf(" for best tree from trial %d", Best);
	printf("\n");
	SoftenThresh(Pruned[Best]);
	printf("\n");
    printf("這個是經過Quinlan剪枝過的");
	PrintTree(Pruned[Best]);
    }

    /*  Save best tree  */

    if ( BATCH || TRIALS == 1 )
    {
	printf("\nTree saved\n");
    }
    else
    {
	printf("\nBest tree from trial %d saved\n", Best);
    }
    SaveTree(Pruned[Best], ".tree");

    /*  Evaluation  */

    printf("\n\nEvaluation on training data (%d items):\n", MaxItem+1);
    Evaluate(false, Best);

    if ( UNSEENS )
    {   
        GetData(".test");
        printf("\nEvaluation on test data (%d items):\n", MaxItem+1);
        Evaluate(true, Best);
    }

    exit(0);
}
